client/tests/phpunit/Form/Admin/ReportChecklistTypeTest.php

```diff
@@ -0,0 +1,101 @@
+{
+
+    /**
+     * @dataProvider formValuesProvider
+     */
+    public function testSubmitValidData(?string $costValues, bool $deputyDetails)
+    {
+        $report = new Report();
+        $report->setAvailableSections(['profDeputyCosts']);
+
+        $formDataObject = new Checklist($report);
+
+        $options = [
+            'report' => $report
+        ];
+
+        $form = $this->factory->create(ReportChecklistType::class, $formDataObject, $options);
+
+        $expected = new Checklist($report);
+        $expected->setPaymentsMatchCostCertificate($costValues);
+        $expected->setProfCostsReasonableAndProportionate($costValues);
+        $expected->setHasDeputyOverchargedFromPreviousEstimates($costValues);
+        $expected->setContactDetailsUptoDate($deputyDetails);
+        $expected->setDeputyFullNameAccurateInCasrec($deputyDetails);
+
+        $formData = [
+            'paymentsMatchCostCertificate' => $costValues,
+            'profCostsReasonableAndProportionate' => $costValues,
+            'hasDeputyOverchargedFromPreviousEstimates' => $costValues,
+        ];
+
+        $form->submit($formData);
+
+        $this->assertTrue($form->isSynchronized());
+
+        $this->assertEquals($expected, $formDataObject);
+    }
+
+    public function formValuesProvider()
+    {
+        return [
+            'Yes' => ['yes', false],
+            'No' => ['no', false],
+            'Null' => [null, false]
+        ];
+    }
+
+    /**
+     * @dataProvider invalidFormValuesProvider
+     */
+    public function testSubmitInvalidData($costCertificates, $reasonableCosts, $overChargedCosts, bool $deputyDetails)
+    {
+        $report = new Report();
+        $report->setAvailableSections(['profDeputyCosts']);
+
+        $formDataObject = new Checklist($report);
+
+        $options = [
+            'report' => $report
+        ];
+
+        $form = $this->factory->create(ReportChecklistType::class, $formDataObject, $options);
+
+        $expected = new Checklist($report);
+        $expected->setPaymentsMatchCostCertificate(null);
+        $expected->setProfCostsReasonableAndProportionate(null);
+        $expected->setHasDeputyOverchargedFromPreviousEstimates(null);
+        $expected->setContactDetailsUptoDate($deputyDetails);
+        $expected->setDeputyFullNameAccurateInCasrec($deputyDetails);
+
+        $formData = [
+            'paymentsMatchCostCertificate' => $costCertificates,
+            'profCostsReasonableAndProportionate' => $reasonableCosts,
+            'hasDeputyOverchargedFromPreviousEstimates' => $overChargedCosts,
+        ];
+
+        $form->submit($formData);
+
+        $this->assertTrue($form->isSynchronized());
+
+        $this->assertEquals($expected, $formDataObject);
+    }
+
+    public function invalidFormValuesProvider()
+    {
+        return [
+            'String too long' => ['abcd','abcd','abcd', false],
+            'Incorrect type' => [1, 2, 3, false]
+        ];
+    }
+}
```

Comment:
I really like bringing form testing into unit tests, although I notice the link you provide mentions to _not_ unit test the form validation:

_Don't test the validation: it is applied by a listener that is not active in the test case and it relies on validation configuration. Instead, unit test your custom constraints directly or read how to add custom extensions in the last section of this page._

So essentially we should be using this to test the mapping between form fields and the Entity, and for the presence of any custom vars that we pass to it. I think the idea is that we only test our custom validation constraints, and trust the Symfony ones - although that does not test that we have configured the validation correctly which is rather important in my opinion. Perhaps we'll have to use behat for form validation after all?

Reply:
Yeah I did notice that but like you said it felt wrong to not test that side of the form...I guess if we wanted to ensure the validation on the entity was configured correctly then a simple direct entity test would prove that by passing in some valid and invalid values to the setters and trust that Symfony internals are doing the right thing.

Reply:
I don't think that will cover the specific constraints that we configure in the annotation of each property in the entity. To do that it looks like we would need to use the `Validation` component in isolation: https://stackoverflow.com/questions/7004601/symfony2-testing-entity-validation-constraints

Reply:
Nice find - I'll see how that goes applying to this ticket and if it works then lets use this approach going forward 👍